# BBM409 Assignment 2 - Detection of Fake News

## Part 1 - Theory Questions

### MLE

#### Question1:
Assume that you have a data consisting of $ x_1 $ , $ x_2 $ ,..., $ x_m $  where each $ x_i $ represents
a single real value, which means you have m instances in data and each instance
has an single real-valued attribute. Assume also that the given data has random
uniform distribution between $ −w $ and $ w $. You are expected to find the maximum
likelihood estimate of w with respect to the given data.

$ * $ Specify a likelihood function $F(w)$  <br />  <br />
$F(w) = \prod_{i=1}^{m}P_w(x_i) $

It is uniformly distributed

$P_w(x_i) =  \Bigg\{ \begin{array}{c}
\frac{1}{2w},\quad \quad if x\in[-w,w], \\
0,\quad \text{otherwise.} \\
\end{array}$

$ * $ Specify the maximum likelihood estimate for w. Consider your answer based on the likelihood function you state.

$ F(w) = \prod_{i=1}^{m}P_w(x_i) = \prod_{i=1}^{m}\frac{1}{2w} = \frac{1}{2w^{m}} = 2w^{-m} $

We carry out ln function both side

$InF(w) = -mIn2w$

$\frac{d \mathbf{InF(w)}}{d \mathbf{w}} = \frac{-m}{2w} $

The result is less than zero when w is bigger than zero

$F(w)$ is a decreasing function in bounds and otherwise it will be zero. This gives,

$ w = max(| min_{1≤i≤m} x_i|, | max_{1≤i≤m} x_i|)$

$ * $ Assume that this time you are given a labelled data $(x_i, y_i)$, where $y_i$ is 1 or 0. Remember that a generative classifier will try to model $P(y)$ and $P(x|y)$. Define an example dataset the generative classifier utilizing the model you defined above for each $P(x|y)$ could not perform well on.

Lets say there are x positive values in range $ [0, N] $ and x negative values in range $ [-N, 0] $. The Maximum Likelihood Estimate will give the value M for w. So we will classify them based on previous classes. The issue is the bias of the classifier. it is not symmetric.

-Remember that a discriminative classifier will try to model P(y|x). State that whether you can classify the labelled data given in previously part using such a discriminative classifier or not. If your answer is answer is yes, then please also show that what your suggested classifier looks like.

Yes. Threshold value can be selected. The value of x can be classified according to whether < or > case.

#### Question 2:
Fill the blanks with T (True) or F (False) for the statements given above.

$ * $ Maximum likelihood estimation provides not only point estimation, but a distribution information of the parameters estimated. (F)

$ * $ Maximum likelihood and Bayesian approachs for parameter estimation perform well with low-dimensional dataset with many training examples while their performance is bad on high-dimensional dataset with few training examples. (T)

### Naive Bayes

#### Question 1:
Consider that you are given the dataset in the table above consisting of boolean variables x, y and z and a single boolean output variable C. Suppose that the Naive Bayes classifier is going to be used.

$ * $ Specify the value of $P(C = 1|x = 1, y = 1, z = 0)$. Show your solution step by step.

$ P(C = 1|x = 1, y = 1, z = 0) = \frac{P(C = 1,x = 1, y = 1, z = 0)} {P(x = 1, y = 1, z = 0)} = \frac{P(C=1) * P(x = 1|C = 1) * P(y = 1|C = 1) * P(z = 0|C = 1)}{P(x = 1 , y = 1 , z = 0 , C = 1) + P(x = 1 , y = 1 , z = 0 , C = 0)} = \frac{1}{4}$

$ * $ Specify the value of $P(C = 0|x = 1, y = 1)$. Show your solution step by step. 

$P(C = 0|x = 1, y = 1) = \frac {P(C = 0 , x = 1 , y = 1)}{P(x = 1 , y = 1)} = \frac {P(C= 0) * P(x = 1|C = 0) * P(y = 1|C = 0)}{P(x = 1 , y = 1 , C = 0) + P(x = 1 , y = 1 , C = 1)} = \frac{1}{2}$

Now suppose that the Joint Naive Bayes classifier is used for the options below:

$ * $ Specify the value of $P(C = 1|x = 1, y = 1, z = 0)$. Show your solution step by step.

$P(C = 1|x = 1, y = 1, z = 0) = \frac{num(C = 1, x = 1, y = 1, z = 0)}{num(x = 1, y = 1, z = 0)}$

$num(C = 1, x = 1, y = 1, z = 0) = 0 $

$P(C = 1|x = 1, y = 1, z = 0) = 0 $

$ * $ Specify the value of $ P(C = 0|x = 1, y = 1)$. Show your solution step by step.

$ P(C = 0|x = 1, y = 1) = \frac{num(C = 0, x = 1, y = 1)}{num(x = 1, y = 1)}$

$ P(C = 0|x = 1, y = 1) = \frac{1}{2} $

#### Question 2:
Assume that you have three variables, which are A, B and C.

$ * $ Suppose that you have the following informations $P(C|A) = 0.7$ and $P(C|B) = 0.4$. State that whether you can compute $P(C|A, B)$ with the informations given previously or not. Besides show your solution if you can and explain the reason if you can not. 

$P(C|A, B) = \frac {P(C,A,B)}{P(A,B)} = \frac {P(C)*P(A|C)*P(B|C)}{P(A,B)}$

$P(A|C) = \frac {P(C|A)*P(A)}{P(C)}$

$P(B|C) = \frac {P(C|B)*P(B)}{P(C)}$

We need the information of $P(A),P(B),P(C)$.

$ * $ Suppose that besides two informations above, $P(A) = 0.3$ and $P(B) = 0.5$ informations are given. State that whether you can compute $P(C|A, B)$ with the informations given previously or not. Besides show your solution if you can and explain the reason if you can not.

$P(C|A, B) = \frac {P(C,A,B)}{P(A,B)} = \frac {P(C)*P(A|C)*P(B|C)}{P(A,B)}$

$P(A|C) = \frac {P(C|A)*P(A)}{P(C)} \Longrightarrow \frac {0.21}{P(C)}$

$P(B|C) = \frac {P(C|B)*P(B)}{P(C)}  \Longrightarrow \frac {0.20}{P(C)}$

$P(A,B)= P(A|B)*P(B) = P(B|A)*P(A)$

We need the information of $P(A|B), P(B|A)$.

$ * $ Finally assume that you have only informations, which are $P(C, A) = 0.2$, $P(A) = 0.3$ and $P(B) = 1$. State that whether you can compute $P(C|A, B)$ with the informations given previously or not. Besides show your solution if you can and explain the reason if you can not.

$P(C|A, B) = \frac {P(C,A,B)}{P(A,B)} = \frac {P(C)*P(A|C)*P(B|C)}{P(A,B)}$

$P(A,B) = P(A) $ in case of $P(B)=1$

$P(B|C) = \frac {P(C|B)*P(B)}{P(C)} = \frac {P(C) * 1  }{P(C)} = P(B|C) = 1 $

$P(C|A, B) = \frac {P(C)*P(A|C)}{P(A)}) = \frac {P(A,C)}{P(A)} = \frac{0.2} {0.3} = \frac {2}{3}$

## Part 2 - Implementation

In [1]:
import pandas as pd
import numpy as np
from random import choice
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

In [2]:
# This function calculates the likelihood probability using laplace smoothing method to handle the situation of the word never presence in text
def calc_cond_prob(numOfOccurences, numOfWords, uniqueWords):
    return float(numOfOccurences + 1) / (numOfWords + uniqueWords)

In [3]:
# This fuction calculates naive bayes probabilities of test news and returns a list of predictions
# It also implement log probability method to prevent numerical underflow
def naive_bayes(test_news_vectorizer, countWordInTestNews, cond_probs_fake, cond_probs_real, fakeWords,
                realWords, length_fake, length_real, uniqueWords, totalFakeProb, totalRealProb):
    labels_of_test = list()
    content_of_test_element = test_news_vectorizer.inverse_transform(countWordInTestNews)
    
    # for fake news probability
    listOfFakeProbs = list()
    for i in range(len(content_of_test_element)):   # for each line
        fakeProb = 1.0
        for j in content_of_test_element[i]:   # for each word in line
            if j in fakeWords:                 # if it is a learned word
                fakeProb *= cond_probs_fake[j]
            else:
                fakeProb *= calc_cond_prob(0, length_fake, uniqueWords)
        fakeProb *= totalFakeProb
        listOfFakeProbs.append(np.log(fakeProb))

    # for real news probability
    listOfRealProbs = list()
    for i in range(len(content_of_test_element)):   # for each line
        realProb = 1.0
        for j in content_of_test_element[i]:     # for each word in line
            if j in realWords:                   # if it is a learned word
                realProb *= cond_probs_real[j]
            else:
                realProb *= calc_cond_prob(0, length_real, uniqueWords)
                realProb *= totalRealProb
        listOfRealProbs.append(np.log(realProb))

    for i in range(len(listOfFakeProbs)):
        labels_of_test.append(1 if listOfFakeProbs[i] > listOfRealProbs[i] else 0)

    return labels_of_test

In [4]:
# calculating the conditional probability, it returns a dictionary whose keys are words and values are probability
def cond_prob(words, totalCountsOfWords, length, uniqueWordsNumber):
    cond_probs = dict()
    for i in range(len(words)):
        word = words[i]
        cond_probs[word] = calc_cond_prob(totalCountsOfWords[i], length, uniqueWordsNumber)
    return cond_probs

In [5]:
# In this code part, dataframe is read from file and splitted to train and test data by ratio of 70%
data = pd.read_csv("fake_news_train.csv")
train_news = data[:round(len(data)*0.7)]
test_news = data[round(len(data)*0.7):]

In [6]:
# I handled the NaN value by changing the absent parts with relevant message
train_news.iloc[:,1].fillna("no_title", inplace=True)
train_news.iloc[:,2].fillna("no_author", inplace=True)
train_news.iloc[:,3].fillna("no_text", inplace=True)

test_news.iloc[:,1].fillna("no_title", inplace=True)
test_news.iloc[:,2].fillna("no_author", inplace=True)
test_news.iloc[:,3].fillna("no_text", inplace=True)

In [7]:
# this part splits the data to fake and real news
fake_news = train_news[train_news['label'] == 1]
real_news = train_news[train_news['label'] == 0]

In [8]:
# This part creates a matrix whose rows hold the news and columns hold the number of occurences using CountVectorizer function
# It stores the information only from title column
train_news_vectorizer = CountVectorizer()
newscv = train_news_vectorizer.fit_transform(train_news['title'])
countWordInTrainNews = newscv.toarray()

test_news_vectorizer = CountVectorizer()
test_newscv = test_news_vectorizer.fit_transform(test_news['title'])
countWordInTestNews = test_newscv.toarray()

fake_vectorizer = CountVectorizer()
fakeNewsCv = fake_vectorizer.fit_transform(fake_news['title'])
countWordInFake = fakeNewsCv.toarray()

real_vectorizer = CountVectorizer()
realNewscv = real_vectorizer.fit_transform(real_news['title'])
countWordInReal = realNewscv.toarray()

In [9]:
# These variables holds the word data in the order of counts of names
train_newsWords = train_news_vectorizer.get_feature_names()
test_newsWords = test_news_vectorizer.get_feature_names()
fakeWords = fake_vectorizer.get_feature_names()
realWords = real_vectorizer.get_feature_names()

In [10]:
# This part creates a structure that holds the number of occurences each word in entire data 
totalCountsOfWordsInTrainNews = np.sum(countWordInTrainNews, axis=0)
totalCountsOfWordsInTestNews = np.sum(countWordInTestNews, axis=0)
totalCountsOfWordsInFake = np.sum(countWordInFake, axis=0)
totalCountsOfWordsInReal = np.sum(countWordInReal, axis=0)

#### Part 1
Next line part analyze the data

In [11]:
# Reorganizing the data and creates the list of words whose presence are only in real or fake words
fakeWordSet = set(fakeWords)
realWordSet = set(realWords)
wordsOnlyReal = realWordSet - fakeWordSet
wordsOnlyFake = fakeWordSet - realWordSet

# Analyzing words and finding the words that is pure to a label, also calculating how many times it occurs
i = 0
while i < 3:
    word = choice(list(wordsOnlyReal))
    indexNum = realWords.index(word)
    if totalCountsOfWordsInReal[indexNum] >= 5:
        print("The word -", word, "- is", totalCountsOfWordsInReal[indexNum],
              "times in real news. Also it is never in fake news")
        i += 1
i = 0
while i < 3:
    word = choice(list(wordsOnlyFake))
    indexNum = fakeWords.index(word)
    if totalCountsOfWordsInFake[indexNum] >= 5:
        print("The word -", word, "- is", totalCountsOfWordsInFake[indexNum],
              "times in fake news. Also it is never in real news")
        i += 1

The word - expands - is 5 times in real news. Also it is never in fake news
The word - cotton - is 6 times in real news. Also it is never in fake news
The word - criticizes - is 5 times in real news. Also it is never in fake news
The word - incredible - is 6 times in fake news. Also it is never in real news
The word - totally - is 10 times in fake news. Also it is never in real news
The word - raqqa - is 13 times in fake news. Also it is never in real news


In [12]:
# This part creates a matrix whose rows hold the news and columns hold the number of occurences using CountVectorizer function
# It takes data from text part and uses min_df and max_df parameter to crop some of data to be able to make program efficient in running time.
unigram_train_news_vectorizer = CountVectorizer(min_df=0.025, max_df=0.975)
unigram_newscv = unigram_train_news_vectorizer.fit_transform(train_news['text'])
unigram_countWordInTrainNews = unigram_newscv.toarray()

unigram_test_news_vectorizer = CountVectorizer(min_df=0.025, max_df=0.975)
unigram_test_newscv = unigram_test_news_vectorizer.fit_transform(test_news['text'])
unigram_countWordInTestNews = unigram_test_newscv.toarray()

unigram_fake_vectorizer = CountVectorizer(min_df=0.025, max_df=0.975)
unigram_fakeNewsCv = unigram_fake_vectorizer.fit_transform(fake_news['text'])
unigram_countWordInFake = unigram_fakeNewsCv.toarray()

unigram_real_vectorizer = CountVectorizer(min_df=0.025, max_df=0.975)
unigram_realNewscv = unigram_real_vectorizer.fit_transform(real_news['text'])
unigram_countWordInReal = unigram_realNewscv.toarray()

In [13]:
# These variables holds the word data in the order of counts of names
unigram_train_newsWords = unigram_train_news_vectorizer.get_feature_names()
unigram_test_newsWords = unigram_test_news_vectorizer.get_feature_names()
unigram_fakeWords = unigram_fake_vectorizer.get_feature_names()
unigram_realWords = unigram_real_vectorizer.get_feature_names()

In [14]:
# This part creates a structure that holds the number of occurences each word in entire data 
unigram_totalCountsOfWordsInTrainNews = np.sum(unigram_countWordInTrainNews, axis=0)
unigram_totalCountsOfWordsInTestNews = np.sum(unigram_countWordInTestNews, axis=0)
unigram_totalCountsOfWordsInFake = np.sum(unigram_countWordInFake, axis=0)
unigram_totalCountsOfWordsInReal = np.sum(unigram_countWordInReal, axis=0)

In [15]:
# This parts calculates P(Fake) and P(Real)
totalFakeProb = fake_news.shape[0] / train_news.shape[0]
totalRealProb = real_news.shape[0] / train_news.shape[0]

In [16]:
# Number of words in a label
unigram_length_fake, unigram_length_real = 0, 0
for i in range(len(unigram_fakeWords)):
    unigram_length_fake += unigram_totalCountsOfWordsInFake[i]

for i in range(len(unigram_realWords)):
    unigram_length_real += unigram_totalCountsOfWordsInReal[i]

It is feasible to predict whether a headline is real or fake news from words that
appear in the headline.

#### Part 2
Implementing Naive Bayes with unigram and bigram
CountVectorizer is set in unigram

In [17]:
# Calculates conditional probabilities for fake and real news apartly
unigram_cond_probs_fake = cond_prob(unigram_fakeWords, unigram_totalCountsOfWordsInFake, unigram_length_fake, len(unigram_train_newsWords))
unigram_cond_probs_real = cond_prob(unigram_realWords, unigram_totalCountsOfWordsInReal, unigram_length_real, len(unigram_train_newsWords))

In [18]:
# Carrying out the Naive Bayes Formula
unigram_labels_of_test = naive_bayes(unigram_test_news_vectorizer, unigram_countWordInTestNews, unigram_cond_probs_fake, unigram_cond_probs_real, unigram_fakeWords, unigram_realWords,
            unigram_length_fake, unigram_length_real, len(unigram_train_newsWords), totalFakeProb, totalRealProb)

C:\Users\Yusuf Emre\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: RuntimeWarning: divide by zero encountered in log
C:\Users\Yusuf Emre\Anaconda3\lib\site-packages\ipykernel_launcher.py:30: RuntimeWarning: divide by zero encountered in log


In [19]:
# Calculating the Accuracy of unigram part
numOfTruePrediction = 0
for i in range(len(unigram_labels_of_test)):
    if test_news.iloc[i, 4] == unigram_labels_of_test[i]:
        numOfTruePrediction += 1
unigram_accuracy = 100*(numOfTruePrediction/float(len(unigram_labels_of_test)))
print(unigram_accuracy)

62.379807692307686


Implementing bigram

In [20]:
# This part creates a matrix whose rows hold the news and columns hold the number of occurences using CountVectorizer function in bigram form
# It takes data from text part and uses min_df and max_df parameter to crop some of data to be able to make program efficient in running time.
bigram_train_news_vectorizer = CountVectorizer(ngram_range=(2,2), min_df=0.025, max_df=0.975)
bigram_newscv = bigram_train_news_vectorizer.fit_transform(train_news['text'])
bigram_countWordInTrainNews = bigram_newscv.toarray()

bigram_test_news_vectorizer = CountVectorizer(ngram_range=(2,2), min_df=0.025, max_df=0.975)
bigram_test_newscv = bigram_test_news_vectorizer.fit_transform(test_news['text'])
bigram_countWordInTestNews = bigram_test_newscv.toarray()

bigram_fake_vectorizer = CountVectorizer(ngram_range=(2,2), min_df=0.025, max_df=0.975)
bigram_fakeNewsCv = bigram_fake_vectorizer.fit_transform(fake_news['text'])
bigram_countWordInFake = bigram_fakeNewsCv.toarray()

bigram_real_vectorizer = CountVectorizer(ngram_range=(2,2), min_df=0.025, max_df=0.975)
bigram_realNewscv = bigram_real_vectorizer.fit_transform(real_news['text'])
bigram_countWordInReal = bigram_realNewscv.toarray()

In [21]:
# These variables holds the word data in the order of counts of names in bigram form
bigram_train_newsWords = bigram_train_news_vectorizer.get_feature_names()
bigram_test_newsWords = bigram_test_news_vectorizer.get_feature_names()
bigram_fakeWords = bigram_fake_vectorizer.get_feature_names()
bigram_realWords = bigram_real_vectorizer.get_feature_names()

In [22]:
# This part creates a structure that holds the number of occurences each word in entire data in bigram form
bigram_totalCountsOfWordsInTrainNews = np.sum(bigram_countWordInTrainNews, axis=0)
bigram_totalCountsOfWordsInTestNews = np.sum(bigram_countWordInTestNews, axis=0)
bigram_totalCountsOfWordsInFake = np.sum(bigram_countWordInFake, axis=0)
bigram_totalCountsOfWordsInReal = np.sum(bigram_countWordInReal, axis=0)

In [23]:
# Number of words in a label in bigram form
bigram_length_fake, bigram_length_real = 0, 0
for i in range(len(bigram_fakeWords)):
    bigram_length_fake += bigram_totalCountsOfWordsInFake[i]

for i in range(len(bigram_realWords)):
    bigram_length_real += bigram_totalCountsOfWordsInReal[i]

In [24]:
# Calculates conditional probabilities for fake and real news apartly
bigram_cond_probs_fake = cond_prob(bigram_fakeWords, bigram_totalCountsOfWordsInFake, bigram_length_fake, len(bigram_train_newsWords))
bigram_cond_probs_real = cond_prob(bigram_realWords, bigram_totalCountsOfWordsInReal, bigram_length_real, len(bigram_train_newsWords))

In [25]:
# Carrying out the Naive Bayes Formula
labels_of_test = naive_bayes(bigram_test_news_vectorizer, bigram_countWordInTestNews, bigram_cond_probs_fake, bigram_cond_probs_real, bigram_fakeWords, bigram_realWords,
            bigram_length_fake, bigram_length_real, len(bigram_train_newsWords), totalFakeProb, totalRealProb)

C:\Users\Yusuf Emre\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: RuntimeWarning: divide by zero encountered in log
C:\Users\Yusuf Emre\Anaconda3\lib\site-packages\ipykernel_launcher.py:30: RuntimeWarning: divide by zero encountered in log


In [26]:
# Calculating the Accuracy of bigram part
numOfTruePrediction = 0
for i in range(len(labels_of_test)):
    if test_news.iloc[i, 4] == labels_of_test[i]:
        numOfTruePrediction += 1
bigram_accuracy = 100*(numOfTruePrediction/float(len(labels_of_test)))
print(bigram_accuracy)

66.08573717948718


#### Part 3
Analyzing effect of the words on prediction

In [27]:
# In  this part, I used TFIDF to take the the weights of the words. In other words, it returns the importance values of the words.
tfidf_train_news_vectorizer = TfidfTransformer()
tfidf_newscv = tfidf_train_news_vectorizer.fit_transform(unigram_newscv)
tfidf_countWordInTrainNews = tfidf_newscv.toarray()

tfidf_test_news_vectorizer = TfidfTransformer()
tfidf_test_newscv = tfidf_test_news_vectorizer.fit_transform(unigram_test_newscv)
tfidf_countWordInTestNews = tfidf_test_newscv.toarray()

tfidf_fake_vectorizer = TfidfTransformer()
tfidf_fakeNewsCv = tfidf_fake_vectorizer.fit_transform(unigram_fakeNewsCv)
tfidf_countWordInFake = tfidf_fakeNewsCv.toarray()

tfidf_real_vectorizer = TfidfTransformer()
tfidf_realNewscv = tfidf_real_vectorizer.fit_transform(unigram_realNewscv)
tfidf_countWordInReal = tfidf_realNewscv.toarray()

In [28]:
tfidf_totalCountsOfWordsInTrainNews = np.sum(tfidf_countWordInTrainNews, axis=0)
tfidf_totalCountsOfWordsInTestNews = np.sum(tfidf_countWordInTestNews, axis=0)
tfidf_totalCountsOfWordsInFake = np.sum(tfidf_countWordInFake, axis=0)
tfidf_totalCountsOfWordsInReal = np.sum(tfidf_countWordInReal, axis=0)

In [29]:
# The words whose presence has more weight to predict correctly in real news help us classifying.
# Likewise, the words whose presence has less weight to predict correctly in fake news help us avoiding misclassification.

indices_sorted_real_news = np.argsort(tfidf_totalCountsOfWordsInReal)
indices_sorted_fake_news = np.argsort(tfidf_totalCountsOfWordsInFake)

print("List of 10 words whose presence most strongly predicts that the news is real.")
for i in indices_sorted_real_news[-10:]:
    if realWords[i] == 'no_title':
        print(realWords[indices_sorted_real_news[-11]])
    print(realWords[i])
    
print("List of 10 words whose absence most strongly predicts that the news is real.")
for i in indices_sorted_real_news[:10]:
    if realWords[i] == 'no_title':
        print(realWords[indices_sorted_real_news[11]])
    print(realWords[i])
    
print("List of 10 words whose presence most strongly predicts that the news is fake.")
for i in indices_sorted_fake_news[-10:]:
    if fakeWords[i] == 'no_title':
        print(fakeWords[indices_sorted_fake_news[-11]])
    print(fakeWords[i])
    
print("List of 10 words whose absence most strongly predicts that the news is fake.")
for i in indices_sorted_fake_news[:10]:
    if fakeWords[i] == 'no_title':
        print(fakeWords[indices_sorted_fake_news[11]])
    print(fakeWords[i])

List of 10 words whose presence most strongly predicts that the news is real.
defecting
curtailed
deaths
booming
commish
booed
bias
battered
cajole
crist
List of 10 words whose absence most strongly predicts that the news is real.
cheers
aussie
ancient
classroom
astronauts
cotton
contributed
answers
costing
backfire
List of 10 words whose presence most strongly predicts that the news is fake.
bombing
bannock
ascension
banker
civilizations
backup
3d
bodyguard
clever
ck
List of 10 words whose absence most strongly predicts that the news is fake.
borders
1975
bastion
arcturian
cage
boris
aids
balistique
arabie
antarctic


In [30]:
tfidf_length_fake, tfidf_length_real = 0, 0
for i in range(tfidf_countWordInFake.shape[1]):
    tfidf_length_fake += tfidf_totalCountsOfWordsInFake[i]

for i in range(tfidf_countWordInReal.shape[1]):
    tfidf_length_real += tfidf_totalCountsOfWordsInReal[i]

In [31]:
tfidf_cond_probs_fake = cond_prob(unigram_fakeWords, tfidf_totalCountsOfWordsInFake, tfidf_length_fake, tfidf_countWordInTrainNews.shape[1])
tfidf_cond_probs_real = cond_prob(unigram_realWords, tfidf_totalCountsOfWordsInReal, tfidf_length_real, tfidf_countWordInTrainNews.shape[1])

In [32]:
labels_of_test = naive_bayes(test_news_vectorizer, tfidf_countWordInTestNews, tfidf_cond_probs_fake, tfidf_cond_probs_real, unigram_fakeWords, unigram_realWords,
            tfidf_length_fake, tfidf_length_real, tfidf_countWordInTrainNews.shape[1], totalFakeProb, totalRealProb)

C:\Users\Yusuf Emre\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: RuntimeWarning: divide by zero encountered in log
C:\Users\Yusuf Emre\Anaconda3\lib\site-packages\ipykernel_launcher.py:30: RuntimeWarning: divide by zero encountered in log


In [33]:
# Calculating the Accuracy of tfidf part
numOfTruePrediction = 0
for i in range(len(labels_of_test)):
    if test_news.iloc[i, 4] == labels_of_test[i]:
        numOfTruePrediction += 1
tfidf_accuracy = 100*(numOfTruePrediction/float(len(labels_of_test)))
print(tfidf_accuracy)

59.71554487179487


**Implementation part 2 with only title data**

Unigram part

In [34]:
# This part creates a matrix whose rows hold the news and columns hold the number of occurences using CountVectorizer function
train_news_vectorizer = CountVectorizer()
newscv = train_news_vectorizer.fit_transform(train_news['title'])
countWordInTrainNews = newscv.toarray()

test_news_vectorizer = CountVectorizer()
test_newscv = test_news_vectorizer.fit_transform(test_news['title'])
countWordInTestNews = test_newscv.toarray()

fake_vectorizer = CountVectorizer()
fakeNewsCv = fake_vectorizer.fit_transform(fake_news['title'])
countWordInFake = fakeNewsCv.toarray()

real_vectorizer = CountVectorizer()
realNewscv = real_vectorizer.fit_transform(real_news['title'])
countWordInReal = realNewscv.toarray()

In [35]:
# These variables holds the word data in the order of counts of names
train_newsWords = train_news_vectorizer.get_feature_names()
test_newsWords = test_news_vectorizer.get_feature_names()
fakeWords = fake_vectorizer.get_feature_names()
realWords = real_vectorizer.get_feature_names()

In [36]:
# This part creates a structure that holds the number of occurences each word in entire data 
totalCountsOfWordsInTrainNews = np.sum(countWordInTrainNews, axis=0)
totalCountsOfWordsInTestNews = np.sum(countWordInTestNews, axis=0)
totalCountsOfWordsInFake = np.sum(countWordInFake, axis=0)
totalCountsOfWordsInReal = np.sum(countWordInReal, axis=0)

In [37]:
# Number of words in a label
length_fake, length_real = 0, 0
for i in range(len(fakeWords)):
    length_fake += totalCountsOfWordsInFake[i]

for i in range(len(realWords)):
    length_real += totalCountsOfWordsInReal[i]

In [38]:
cond_probs_fake = cond_prob(fakeWords, totalCountsOfWordsInFake, length_fake, len(train_newsWords))
cond_probs_real = cond_prob(realWords, totalCountsOfWordsInReal, length_real, len(train_newsWords))

In [39]:
labels_of_test = naive_bayes(test_news_vectorizer, countWordInTestNews, cond_probs_fake, cond_probs_real, fakeWords, realWords,
            length_fake, length_real, len(train_newsWords), totalFakeProb, totalRealProb)

In [40]:
# Calculating the Accuracy of unigram part
numOfTruePrediction = 0
for i in range(len(labels_of_test)):
    if test_news.iloc[i, 4] == labels_of_test[i]:
        numOfTruePrediction += 1
unigram_accuracy = 100*(numOfTruePrediction/float(len(labels_of_test)))
print(unigram_accuracy)

89.90384615384616


Bigram part

In [41]:
# This part creates a matrix whose rows hold the news and columns hold the number of occurences using CountVectorizer function in bigram form
bigram_train_news_vectorizer = CountVectorizer(ngram_range=(2,2))
bigram_newscv = bigram_train_news_vectorizer.fit_transform(train_news['title'])
bigram_countWordInTrainNews = bigram_newscv.toarray()

bigram_test_news_vectorizer = CountVectorizer(ngram_range=(2,2))
bigram_test_newscv = bigram_test_news_vectorizer.fit_transform(test_news['title'])
bigram_countWordInTestNews = bigram_test_newscv.toarray()

bigram_fake_vectorizer = CountVectorizer(ngram_range=(2,2))
bigram_fakeNewsCv = bigram_fake_vectorizer.fit_transform(fake_news['title'])
bigram_countWordInFake = bigram_fakeNewsCv.toarray()

bigram_real_vectorizer = CountVectorizer(ngram_range=(2,2))
bigram_realNewscv = bigram_real_vectorizer.fit_transform(real_news['title'])
bigram_countWordInReal = bigram_realNewscv.toarray()

In [42]:
# These variables holds the word data in the order of counts of names in bigram form
bigram_train_newsWords = bigram_train_news_vectorizer.get_feature_names()
bigram_test_newsWords = bigram_test_news_vectorizer.get_feature_names()
bigram_fakeWords = bigram_fake_vectorizer.get_feature_names()
bigram_realWords = bigram_real_vectorizer.get_feature_names()

In [43]:
# This part creates a structure that holds the number of occurences each word in entire data in bigram form
bigram_totalCountsOfWordsInTrainNews = np.sum(bigram_countWordInTrainNews, axis=0)
bigram_totalCountsOfWordsInTestNews = np.sum(bigram_countWordInTestNews, axis=0)
bigram_totalCountsOfWordsInFake = np.sum(bigram_countWordInFake, axis=0)
bigram_totalCountsOfWordsInReal = np.sum(bigram_countWordInReal, axis=0)

In [44]:
# Number of words in a label in bigram form
bigram_length_fake, bigram_length_real = 0, 0
for i in range(len(bigram_fakeWords)):
    bigram_length_fake += bigram_totalCountsOfWordsInFake[i]

for i in range(len(bigram_realWords)):
    bigram_length_real += bigram_totalCountsOfWordsInReal[i]

In [45]:
bigram_cond_probs_fake = cond_prob(bigram_fakeWords, bigram_totalCountsOfWordsInFake, bigram_length_fake, len(bigram_train_newsWords))
bigram_cond_probs_real = cond_prob(bigram_realWords, bigram_totalCountsOfWordsInReal, bigram_length_real, len(bigram_train_newsWords))

In [46]:
labels_of_test = naive_bayes(bigram_test_news_vectorizer, bigram_countWordInTestNews, bigram_cond_probs_fake, bigram_cond_probs_real, bigram_fakeWords, bigram_realWords,
            bigram_length_fake, bigram_length_real, len(bigram_train_newsWords), totalFakeProb, totalRealProb)

In [47]:
# Calculating the Accuracy of bigram part
numOfTruePrediction = 0
for i in range(len(labels_of_test)):
    if test_news.iloc[i, 4] == labels_of_test[i]:
        numOfTruePrediction += 1
bigram_accuracy = 100*(numOfTruePrediction/float(len(labels_of_test)))
print(bigram_accuracy)

79.26682692307693


TfIdf part

In [48]:
# In  this part, I used TFIDF to take the the weights of the words. In other words, it returns the importance values of the words.
tfidf_train_news_vectorizer = TfidfTransformer()
tfidf_newscv = tfidf_train_news_vectorizer.fit_transform(newscv)
tfidf_countWordInTrainNews = tfidf_newscv.toarray()

tfidf_test_news_vectorizer = TfidfTransformer()
tfidf_test_newscv = tfidf_test_news_vectorizer.fit_transform(test_newscv)
tfidf_countWordInTestNews = tfidf_test_newscv.toarray()

tfidf_fake_vectorizer = TfidfTransformer()
tfidf_fakeNewsCv = tfidf_fake_vectorizer.fit_transform(fakeNewsCv)
tfidf_countWordInFake = tfidf_fakeNewsCv.toarray()

tfidf_real_vectorizer = TfidfTransformer()
tfidf_realNewscv = tfidf_real_vectorizer.fit_transform(realNewscv)
tfidf_countWordInReal = tfidf_realNewscv.toarray()

In [49]:
tfidf_totalCountsOfWordsInTrainNews = np.sum(tfidf_countWordInTrainNews, axis=0)
tfidf_totalCountsOfWordsInTestNews = np.sum(tfidf_countWordInTestNews, axis=0)
tfidf_totalCountsOfWordsInFake = np.sum(tfidf_countWordInFake, axis=0)
tfidf_totalCountsOfWordsInReal = np.sum(tfidf_countWordInReal, axis=0)

In [50]:
# The words whose presence has more weight to predict correctly in real news help us classifying.
# Likewise, the words whose presence has less weight to predict correctly in fake news help us avoiding misclassification.

indices_sorted_real_news = np.argsort(tfidf_totalCountsOfWordsInReal)
indices_sorted_fake_news = np.argsort(tfidf_totalCountsOfWordsInFake)

print("List of 10 words whose presence most strongly predicts that the news is real.")
for i in indices_sorted_real_news[-10:]:
    if realWords[i] == 'no_title':
        print(realWords[indices_sorted_real_news[-11]])
    print(realWords[i])
    
print("List of 10 words whose absence most strongly predicts that the news is real.")
for i in indices_sorted_real_news[:10]:
    if realWords[i] == 'no_title':
        print(realWords[indices_sorted_real_news[11]])
    print(realWords[i])
    
print("List of 10 words whose presence most strongly predicts that the news is fake.")
for i in indices_sorted_fake_news[-10:]:
    if fakeWords[i] == 'no_title':
        print(fakeWords[indices_sorted_fake_news[-11]])
    print(fakeWords[i])
    
print("List of 10 words whose absence most strongly predicts that the news is fake.")
for i in indices_sorted_fake_news[:10]:
    if fakeWords[i] == 'no_title':
        print(fakeWords[indices_sorted_fake_news[11]])
    print(fakeWords[i])

List of 10 words whose presence most strongly predicts that the news is real.
for
of
in
breitbart
trump
to
times
york
new
the
List of 10 words whose absence most strongly predicts that the news is real.
infidel
misogynistic
immorality
excerpt
adrienne
opinions
whitlock
nats
strickland
harper
List of 10 words whose presence most strongly predicts that the news is fake.
on
for
hillary
and
trump
in
of
to
the
is
no_title
List of 10 words whose absence most strongly predicts that the news is fake.
rothschild
deviants
nationalistic
ordinanc
ordinance
waking
insatiable
algae
tropical
blooms


In [51]:
tfidf_length_fake, tfidf_length_real = 0, 0
for i in range(tfidf_countWordInFake.shape[1]):
    tfidf_length_fake += tfidf_totalCountsOfWordsInFake[i]

for i in range(tfidf_countWordInReal.shape[1]):
    tfidf_length_real += tfidf_totalCountsOfWordsInReal[i]

In [52]:
tfidf_cond_probs_fake = cond_prob(fakeWords, tfidf_totalCountsOfWordsInFake, tfidf_length_fake, tfidf_countWordInTrainNews.shape[1])
tfidf_cond_probs_real = cond_prob(realWords, tfidf_totalCountsOfWordsInReal, tfidf_length_real, tfidf_countWordInTrainNews.shape[1])

In [53]:
labels_of_test = naive_bayes(test_news_vectorizer, tfidf_countWordInTestNews, tfidf_cond_probs_fake, tfidf_cond_probs_real, fakeWords, realWords,
            tfidf_length_fake, tfidf_length_real, tfidf_countWordInTrainNews.shape[1], totalFakeProb, totalRealProb)

In [54]:
# Calculating the Accuracy of tfidf part
numOfTruePrediction = 0
for i in range(len(labels_of_test)):
    if test_news.iloc[i, 4] == labels_of_test[i]:
        numOfTruePrediction += 1
tfidf_accuracy = 100*(numOfTruePrediction/float(len(labels_of_test)))
print(tfidf_accuracy)

87.21955128205127


Taking into consideration of stopwords


Using stop words speeds up the processing part. Because there will be less feature names

In [55]:
train_news_vectorizer = CountVectorizer(stop_words='english')
newscv = train_news_vectorizer.fit_transform(train_news['title'])

test_news_vectorizer = CountVectorizer(stop_words='english')
test_newscv = test_news_vectorizer.fit_transform(test_news['title'])

fake_vectorizer = CountVectorizer(stop_words='english')
fakeNewsCv = fake_vectorizer.fit_transform(fake_news['title'])

real_vectorizer = CountVectorizer(stop_words='english')
realNewscv = real_vectorizer.fit_transform(real_news['title'])

tfidf_train_news_vectorizer = TfidfTransformer()
tfidf_newscv = tfidf_train_news_vectorizer.fit_transform(newscv)
tfidf_countWordInTrainNews = tfidf_newscv.toarray()

tfidf_test_news_vectorizer = TfidfTransformer()
tfidf_test_newscv = tfidf_test_news_vectorizer.fit_transform(test_newscv)
tfidf_countWordInTestNews = tfidf_test_newscv.toarray()

tfidf_fake_vectorizer = TfidfTransformer()
tfidf_fakeNewsCv = tfidf_fake_vectorizer.fit_transform(fakeNewsCv)
tfidf_countWordInFake = tfidf_fakeNewsCv.toarray()

tfidf_real_vectorizer = TfidfTransformer()
tfidf_realNewscv = tfidf_real_vectorizer.fit_transform(realNewscv)
tfidf_countWordInReal = tfidf_realNewscv.toarray()

In [56]:
tfidf_totalCountsOfWordsInTrainNews = np.sum(tfidf_countWordInTrainNews, axis=0)
tfidf_totalCountsOfWordsInTestNews = np.sum(tfidf_countWordInTestNews, axis=0)
tfidf_totalCountsOfWordsInFake = np.sum(tfidf_countWordInFake, axis=0)
tfidf_totalCountsOfWordsInReal = np.sum(tfidf_countWordInReal, axis=0)

In [57]:
indices_sorted_real_news = np.argsort(tfidf_totalCountsOfWordsInReal)
indices_sorted_fake_news = np.argsort(tfidf_totalCountsOfWordsInFake)

print("List of 10 words whose presence most strongly predicts that the news is real.")
for i in indices_sorted_real_news[-10:]:
    if realWords[i] == 'no_title':
        print(realWords[indices_sorted_real_news[-11]])
    print(realWords[i])
    
print("List of 10 words whose absence most strongly predicts that the news is real.")
for i in indices_sorted_real_news[:10]:
    if realWords[i] == 'no_title':
        print(realWords[indices_sorted_real_news[11]])
    print(realWords[i])
    
print("List of 10 words whose presence most strongly predicts that the news is fake.")
for i in indices_sorted_fake_news[-10:]:
    if fakeWords[i] == 'no_title':
        print(fakeWords[indices_sorted_fake_news[-11]])
    print(fakeWords[i])
    
print("List of 10 words whose absence most strongly predicts that the news is fake.")
for i in indices_sorted_fake_news[:10]:
    if fakeWords[i] == 'no_title':
        print(fakeWords[indices_sorted_fake_news[11]])
    print(fakeWords[i])

List of 10 words whose presence most strongly predicts that the news is real.
rouhani
nests
brandenburg
cked
diversifies
boys
tina
tapper
wellesley
murray
List of 10 words whose absence most strongly predicts that the news is real.
mengele
imdb
every
industry
birther
craftsman
criticism
bratton
satire
editor
List of 10 words whose presence most strongly predicts that the news is fake.
altruistic
fade
richest
ust
mysterious
eaters
civilian
headline
tirol
nathan
List of 10 words whose absence most strongly predicts that the news is fake.
urges
obesity
obese
description
retreat
mother
loch
klux
barbed
canada


#### Part 4
I calculated the accuracy in the ends of all parts.

# Report

In this assignment, I developed a detection program of fake or real news by using probability with a bayesian approach. CountVectorizer and TfidfVectorizer from sklearn library helped so much. This functions saved me from sparse matrixes, and also it counted the words' occurences and frequency of occurences. After obtaining these important data, program calculates probability with naive bayes formula. So it gives the probability of the news is fake/real. Program decides the news label by the most probable label. I also calculated the log probability to prevent numerical underflow, and used Laplace smoothing in case of encountering never occurred words.

$ P(w|c) = \frac{count(w,c) + 1}{count(c) + |V|} $<br />
Plus 1 and |V| came from Laplace smoothing. |V| denotes the unique words in data.

Firstly, when I am trying to understand dataset, I used only title column. I extracted some special words. These words are pure to a label. I printed 3 examples of them.

After understanding dataset, I tried to use text data to predict labels with implementing Naive Bayes. But my program needed so much time to give results and also gave MemoryError to hold that much data. After I consulted my TA, he advised me to give parameters for CountVectorizer. I gave different parameters to CountVectorizer, min_df and max_df. But this parameters caused to lose important part of data to predict correctly. So my predictions' accuracy is pretty low. In unigram part (which evaluate the words apartly from each others) I gave to CountVectorizermin_df=0.025, max_df=0.975 and it gave 62.379807692307686 accuracy. In bigram part (which evaluate the words in group of two words) I gave to CountVectorizer min_df=0.025, max_df=0.975 and it gave 66.08573717948718 accuracy. In tfidf part ( which gives a weighting scheme, that is the importance of words ), it gave 59.71554487179487 accuracy, also when I am finding 10 words whose presence/absence most strongly predicts that the news is fake/real news, I did not encounter with stop words because of min_df and max_df parameters. These accuracy values are so low. Also when I am using text data and calculating naive bayes, interpreter gave ZeroDivision Warning in log operation. I debugged my program and I observed that the warning came from the lines having so many different words. It multiplied the conditional probabilities and the value underflowed.

After these low accuracies, I wanted to try to evaluate with only title data. With title data, I didn't need to ignore the words most and least frequent by using min_df and max_df parameters. In this part, my program gave pretty good accuracy. In unigram part, the accuracy is 89.90384615384616. In bigram part, the accuracy 79.26682692307693. In Tfidf part, the accuracy 87.21955128205127.

When using text data in CountVectorizer, I did not encounter with stop words, because I ignored these terms in min_df and max_df parameters. But to be able to see the effect of stop words I again calculated the "10 words whose presence/absence most strongly predicts that the news is real/fake" in the part where I used only title data. In this part I used the stop_words='english' parameter in CountVectorizer. Using this parameter gave me more exact results for these 10 words.

In last part, I calculated accuracy of all the methods(Unigram, Bigram, TfIdf). These results are right after each part to give more compact representation.

To point a moral, in the part where I take more accurate results, the unigram model gave me more accurate models. I believed that bigram models would give more accuracy in any way, but my program gave more accuracy in unigram. Also TfIdf vectorizer so much helped to extract meaningful information from data.